### PR TITLE
[SCFToCalyx] [StaticLogicToCalyx] Separate lowerings to Calyx.

### PR DESCRIPTION
**Pull request by cgyurgyik**
This adds boilerplate for separating the two conversions to Calyx. 

An additional namespace was appended to each conversion to avoid linking errors; these may be removed once the commonalities are pulled out. I've only pulled out the memory interface for now; the rest will follow. Once these are unglued, we can remove SCF dependencies from the StaticLogic conversion.

#2988

### CODE REVIEW

**Comment by cgyurgyik** [comment]
I removed the macro code in favor of duplicating in https://github.com/llvm/circt/pull/3155/commits/f8baae6c1aa093394a2ccc01d734044221989947, since it is easier to digest. If you're totally against it, I can revert. I personally don't think the amount of duplicated code is large enough to warrant a macro.

**Review by mikeurbach — COMMENTED** [review]
Thanks for starting to separate this out. This is a little different than I had imagined, since all the shared goop is now in StaticLogicToCalyx (e.g. ProgramLoweringState), and it appears gets linked in so SCFToCalyx can still use it. I was imagining all that goop would go into CalyxLoweringUtils.cpp, which both passes could link. This is what I was getting at in step 1. that I outlined in the issue for this.

Do you think that makes sense? I'd hope both SCFToCalyx.cpp and StaticLogicToCalyx.cpp would basically just contain a runOnOperation implementation, and anything that is only needed for each respective pass.

If that's still the end goal but you're working towards it in another angle, then this seems fine, but I'm curious if that's the end goal you have in mind here.

**Comment by cgyurgyik** [comment]
What do you mean it gets linked in? The two files are essentially duplicates, and in the next PR(s) I will pull out all shared functions into the lowering until. 

I believe that's reiterating what you've said. Let me know if I'm missing something.

**Comment by mikeurbach** [comment]
Ah yes sorry, I didn't see that all this stuff is still behind in SCFToCalyx. That sounds good to me, and this PR seems fine after tests pass.

**Comment by cgyurgyik** [comment]
Does this failure look familiar to you @mikeurbach ? I'm not sure what it means by unstaged files, or if it is some clang-tidy related thing.

https://github.com/llvm/circt/runs/6517485399?check_suite_focus=true

**Comment by mikeurbach** [comment]
It's just saying clang tidy failed. We run this script: https://github.com/llvm/circt/blob/68280a14263239ad6a2001d33310decb1fc2f698/.github/workflows/buildAndTest.yml#L337

File: lib/Conversion/StaticLogicToCalyx/StaticLogicToCalyx.cpp [hunk 1]
```diff
@@ -0,0 +1,2742 @@
...
+    return isa<scf::WhileOp, staticlogic::PipelineWhileOp>(op);
+  }
+
+  Block::BlockArgListType getBodyArgs() {
+    return TypeSwitch<Operation *, Block::BlockArgListType>(impl)
+        .Case([](scf::WhileOp op) { return op.getAfterArguments(); })
+        .Case([](staticlogic::PipelineWhileOp op) {
+          return op.getStagesBlock().getArguments();
+        });
+    ;
```
Comment (mikeurbach) [inline lib/Conversion/StaticLogicToCalyx/StaticLogicToCalyx.cpp:76]: extra?

**Comment by mikeurbach** [comment]
Hmm i'm not sure why the NOLINT thing doesn't work. clang-tidy is going to complain about these whenever the code is touched. What do you think about fixing them now in the new code, and when refactoring, use the version with the appropriate lint waivers? If the lint warnings are getting cleaned up anyway, seems worth doing here so the CI passes on this PR and when merged to main. On the other hand, it's not the end of the world if you'd rather wait until the refactoring.

**Comment by cgyurgyik** [comment]
It looks like there's a non-trivial number of complaints, if I'm reading the resulting error correctly. My intention here was to do this incrementally - submit a clean copy in each directory in this CL, pull out the commonalities in the next CL(s). If this is not going to block other folks, then I'd suggest pushing this now so that I can begin the refactoring.

If this will block other folks, then perhaps I need to instead rework this - pull out commonalities (and fix subsequent clang-tidy complaints), *then* separate the two lowerings.

**Comment by mikeurbach** [comment]
> If this will block other folks, then perhaps I need to instead rework this - pull out commonalities (and fix subsequent clang-tidy complaints), then separate the two lowerings.

This is the order I had in mind, but I don't think it's going to block anyone in the meantime, since it would just be you or me working on this.

**Review by mikeurbach — APPROVED** [review]

**Comment by cgyurgyik** [comment]
> > If this will block other folks, then perhaps I need to instead rework this - pull out commonalities (and fix subsequent clang-tidy complaints), then separate the two lowerings.
> 
> This is the order I had in mind, but I don't think it's going to block anyone in the meantime, since it would just be you or me working on this.

That was my initial plan, but I think some pieces may be more cleanly refactored if we have the two classes separated, e.g., actually have interfaces for classes named "interface." Will that actually be the case? We shall find out

**Pull request merged**